src/main/Main.hs

```diff
@@ -626,69 +650,80 @@ execCmd ExecOpts {..} go@GlobalOpts{..} =
     case eoExtra of
         ExecOptsPlain -> do
             (manager,lc) <- liftIO $ loadConfigWithOpts go
-            runStackTGlobal manager (lcConfig lc) go $
+            withUserFileLock (lcConfig lc) $ \lk ->
+             runStackTGlobal manager (lcConfig lc) go $
                 Docker.execWithOptionalContainer
                     (lcProjectRoot lc)
                     (return (eoCmd, eoArgs, [], id))
                     Nothing
-                    (runStackTGlobal manager (lcConfig lc) go $
+                    (runStackTGlobal manager (lcConfig lc) go $ do
+                        liftIO $ unlockFile lk -- Unlock before transferring control away.
                         exec plainEnvSettings eoCmd eoArgs)
                     Nothing
         ExecOptsEmbellished {..} ->
-           withBuildConfig go $ do
+           withBuildConfigAndLock go $ \lk -> do
                let targets = concatMap words eoPackages
                unless (null targets) $
                    Stack.Build.build (const $ return ()) defaultBuildOpts
                        { boptsTargets = map T.pack targets
                        }
+               liftIO $ unlockFile lk -- Unlock before transferring control away.
```

Comment:
Not sure how idiomatic it would be or if it would be much value-add, but `lk` could be an unlocking function instead. E.g. `withBuildConfigAndLock go $ \unlock -> do ..... unlock; exec`.